File: mobile/lib/data/repositories/in_memory_learner_repository.dart
import 'dart:math';
import '../../domain/repositories/i_learner_repository.dart';
import '../../domain/entities/learner_profile.dart';
import '../../domain/entities/skill_score.dart';
import '../../domain/personalization/skill_scoring_engine.dart';

class InMemoryLearnerRepository implements ILearnerRepository {
  final List<AnswerAttemptRecord> _answerAttempts = [];
  final Set<String> _processedActionKeys = {};
  final Set<String> _recordedActivityDates = {};

  LearnerProfile _profile = LearnerProfile(
    id: 'learner_mobile_01',
    name: 'O‘quvchi',
    selectedCourseId: 'course_math_01',
    goal: OnboardingGoal.mastery,
    dailyMinutes: 15,
    initialLevel: InitialLevel.intermediate,
    xp: 0,
    streakDays: 1,
    lastActiveDate: DateTime.now().toIso8601String().split('T').first,
    scoresByCourse: {},
    completedLessonIds: [],
    completedNodeIds: [],
    completedReinforcementIds: [],
    createdAt: DateTime.now().millisecondsSinceEpoch,
  );

  @override
  Future<LearnerProfile> getProfile() async {
    return _profile;
  }

  @override
  Future<LearnerProfile> updateProfile(LearnerProfile updated) async {
    _profile = updated;
    return _profile;
  }

  @override
  Future<void> saveSkillScores(String courseId, Map<String, SkillScore> scores) async {
    final currentScores = Map<String, Map<String, SkillScore>>.from(_profile.scoresByCourse);
    final courseMap = Map<String, SkillScore>.from(currentScores[courseId] ?? {});
    courseMap.addAll(scores);
    currentScores[courseId] = courseMap;

    _profile = _profile.copyWith(scoresByCourse: currentScores);
  }

  @override
  Future<String> savePlacementAttempt(PlacementAttemptData data) async {
    return 'plc_att_${DateTime.now().millisecondsSinceEpoch}';
  }

  @override
  Future<void> markLessonCompleted(String lessonId) async {
    if (!_profile.completedLessonIds.contains(lessonId)) {
      final list = List<String>.from(_profile.completedLessonIds)..add(lessonId);
      _profile = _profile.copyWith(completedLessonIds: list);
      await addXp(20, 'lesson_completed_$lessonId');
      await recordDailyActivity();
    }
  }

  @override
  Future<void> markNodeCompleted(String nodeId) async {
    if (!_profile.completedNodeIds.contains(nodeId)) {
      final list = List<String>.from(_profile.completedNodeIds)..add(nodeId);
      _profile = _profile.copyWith(completedNodeIds: list);
    }
  }

  @override
  Future<void> markReinforcementCompleted(String reinforcementId) async {
    if (!_profile.completedReinforcementIds.contains(reinforcementId)) {
      final list = List<String>.from(_profile.completedReinforcementIds)..add(reinforcementId);
      _profile = _profile.copyWith(completedReinforcementIds: list);
      await addXp(15, 'reinforcement_completed_$reinforcementId');
      await recordDailyActivity();
    }
  }

  @override
  Future<AnswerAttemptRecord> recordAnswerAttempt({
    required String courseId,
    required String skillId,
    required String lessonId,
    required String questionId,
    required int selectedIndex,
    required String selectedAnswer,
    required bool isCorrect,
  }) async {
    final record = AnswerAttemptRecord(
      id: 'att_${DateTime.now().millisecondsSinceEpoch}_${Random().nextInt(9999)}',
      courseId: courseId,
      skillId: skillId,
      lessonId: lessonId,
      questionId: questionId,
      selectedIndex: selectedIndex,
      selectedAnswer: selectedAnswer,
      isCorrect: isCorrect,
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );
    _answerAttempts.add(record);

    if (isCorrect) {
      await addXp(2, 'answer_${questionId}_${record.id}');
    }
    await recordDailyActivity();

    // Recalculate cumulative skill score
    final skillAttempts = _answerAttempts
        .where((a) => a.courseId == courseId && a.skillId == skillId)
        .toList();
    final correctCount = skillAttempts.where((a) => a.isCorrect).length;
    final newScoreVal = SkillScoringEngine.computeSkillScore(correctCount, skillAttempts.length);

    final currentScores = Map<String, Map<String, SkillScore>>.from(_profile.scoresByCourse);
    final courseMap = Map<String, SkillScore>.from(currentScores[courseId] ?? {});
    courseMap[skillId] = SkillScore(
      skillId: skillId,
      courseId: courseId,
      score: newScoreVal,
      lastUpdated: DateTime.now().millisecondsSinceEpoch,
      masteryLevel: SkillScoringEngine.getMasteryLevel(newScoreVal),
    );
    currentScores[courseId] = courseMap;
    _profile = _profile.copyWith(scoresByCourse: currentScores);

    return record;
  }

  @override
  Future<List<AnswerAttemptRecord>> getAnswerAttempts([int limit = 10]) async {
    if (_answerAttempts.length <= limit) {
      return List.from(_answerAttempts);
    }
    return _answerAttempts.sublist(_answerAttempts.length - limit);
  }

  @override
  Future<int> addXp(int amount, [String? actionIdempotencyKey]) async {
    if (actionIdempotencyKey != null) {
      if (_processedActionKeys.contains(actionIdempotencyKey)) {
        return _profile.xp;
      }
      _processedActionKeys.add(actionIdempotencyKey);
    }
    final newXp = _profile.xp + amount;
    _profile = _profile.copyWith(xp: newXp);
    return newXp;
  }

  @override
  Future<int> recordDailyActivity([String? dateStr]) async {
    final today = dateStr ?? DateTime.now().toIso8601String().split('T').first;
    if (_recordedActivityDates.contains(today)) {
      return _profile.streakDays;
    }

    _recordedActivityDates.add(today);
    final lastDate = _profile.lastActiveDate;

    int newStreak = _profile.streakDays;
    if (lastDate.isEmpty) {
      newStreak = 1;
    } else if (lastDate == today) {
      // unchanged
    } else {
      try {
        final prevDt = DateTime.parse(lastDate);
        final currDt = DateTime.parse(today);
        final diffDays = currDt.difference(prevDt).inDays;
        if (diffDays == 1) {
          newStreak = _profile.streakDays + 1;
        } else {
          newStreak = 1;
        }
      } catch (_) {
        newStreak = 1;
      }
    }

    _profile = _profile.copyWith(streakDays: newStreak, lastActiveDate: today);
    return newStreak;
  }

  @override
  Future<LearnerProfile> resetAll() async {
    _profile = LearnerProfile(
      id: 'learner_mobile_01',
      name: 'O‘quvchi',
      selectedCourseId: 'course_math_01',
      goal: OnboardingGoal.mastery,
      dailyMinutes: 15,
      initialLevel: InitialLevel.intermediate,
      xp: 0,
      streakDays: 1,
      lastActiveDate: DateTime.now().toIso8601String().split('T').first,
      scoresByCourse: {},
      completedLessonIds: [],
      completedNodeIds: [],
      completedReinforcementIds: [],
      createdAt: DateTime.now().millisecondsSinceEpoch,
    );
    _answerAttempts.clear();
    _processedActionKeys.clear();
    _recordedActivityDates.clear();
    return _profile;
  }
}

File: src/data/repositories/InMemoryLearnerRepository.ts
import { ILearnerRepository, AnswerAttemptRecord, PlacementAttemptData } from '@/domain/repositories/ILearnerRepository';
import { LearnerProfile } from '@/domain/entities/LearnerProfile';
import { SkillScore } from '@/domain/entities/SkillScore';
import { SkillScoringEngine } from '@/domain/personalization/SkillScoringEngine';

const DEFAULT_PROFILE: LearnerProfile = {
  id: 'learner_default_01',
  name: 'Foydalanuvchi',
  firstName: 'Foydalanuvchi',
  lastName: '',
  email: '',
  selectedCourseId: 'course_math_01',
  goal: 'mastery',
  dailyMinutes: 15,
  initialLevel: 'intermediate',
  xp: 0,
  streakDays: 1,
  lastActiveDate: new Date().toISOString().split('T')[0],
  scoresByCourse: {},
  completedLessonIds: [],
  completedNodeIds: [],
  completedReinforcementIds: [],
  createdAt: Date.now(),
};

export class InMemoryLearnerRepository implements ILearnerRepository {
  private profile: LearnerProfile = { ...DEFAULT_PROFILE };
  private answerAttempts: AnswerAttemptRecord[] = [];
  private processedActionKeys: Set<string> = new Set();
  private recordedActivityDates: Set<string> = new Set();

  async getCurrentProfile(): Promise<LearnerProfile> {
    return this.getProfile();
  }

  async getProfile(): Promise<LearnerProfile> {
    return JSON.parse(JSON.stringify(this.profile));
  }

  async updateProfile(updates: Partial<LearnerProfile>): Promise<LearnerProfile> {
    this.profile = {
      ...this.profile,
      ...updates,
      scoresByCourse: updates.scoresByCourse || this.profile.scoresByCourse,
    };
    return this.getProfile();
  }

  async saveSkillScores(courseId: string, scores: Record<string, SkillScore>): Promise<void> {
    if (!this.profile.scoresByCourse) {
      this.profile.scoresByCourse = {};
    }
    this.profile.scoresByCourse[courseId] = {
      ...(this.profile.scoresByCourse[courseId] || {}),
      ...scores,
    };
  }

  async savePlacementAttempt(_data: PlacementAttemptData): Promise<string> {
    const attemptId = 'plc_att_' + Date.now() + '_' + Math.random().toString(36).substr(2, 5);
    return attemptId;
  }

  async markLessonCompleted(lessonId: string): Promise<void> {
    if (!this.profile.completedLessonIds.includes(lessonId)) {
      this.profile.completedLessonIds.push(lessonId);
      await this.addXp(20, `lesson_completed_${lessonId}`);
      await this.recordDailyActivity();
    }
  }

  async markNodeCompleted(nodeId: string): Promise<void> {
    if (!this.profile.completedNodeIds.includes(nodeId)) {
      this.profile.completedNodeIds.push(nodeId);
    }
  }

  async markReinforcementCompleted(reinforcementId: string): Promise<void> {
    if (!this.profile.completedReinforcementIds.includes(reinforcementId)) {
      this.profile.completedReinforcementIds.push(reinforcementId);
      await this.addXp(15, `reinforcement_completed_${reinforcementId}`);
      await this.recordDailyActivity();
    }
  }

  async recordAnswerAttempt(attempt: Omit<AnswerAttemptRecord, 'id' | 'timestamp'>): Promise<AnswerAttemptRecord> {
    const record: AnswerAttemptRecord = {
      ...attempt,
      id: 'att_' + Date.now() + '_' + Math.random().toString(36).substr(2, 5),
      timestamp: Date.now(),
    };
    this.answerAttempts.push(record);

    // If answer is correct, award +2 XP idempotently
    if (attempt.isCorrect) {
      await this.addXp(2, `answer_${attempt.questionId}_${record.id}`);
    }
    await this.recordDailyActivity();

    // Recalculate cumulative skill score for this skill
    const skillAttempts = this.answerAttempts.filter(
      (a) => a.courseId === attempt.courseId && a.skillId === attempt.skillId
    );
    const correctCount = skillAttempts.filter((a) => a.isCorrect).length;
    const newScoreVal = SkillScoringEngine.computeSkillScore(correctCount, skillAttempts.length);

    if (!this.profile.scoresByCourse[attempt.courseId]) {
      this.profile.scoresByCourse[attempt.courseId] = {};
    }

    this.profile.scoresByCourse[attempt.courseId][attempt.skillId] = {
      skillId: attempt.skillId,
      courseId: attempt.courseId,
      score: newScoreVal,
      lastUpdated: Date.now(),
      masteryLevel: SkillScoringEngine.getMasteryLevel(newScoreVal),
    };

    return record;
  }

  async getAnswerAttempts(limit: number = 10): Promise<AnswerAttemptRecord[]> {
    return [...this.answerAttempts].slice(-limit);
  }

  async addXp(amount: number, actionIdempotencyKey?: string): Promise<number> {
    if (actionIdempotencyKey) {
      if (this.processedActionKeys.has(actionIdempotencyKey)) {
        return this.profile.xp;
      }
      this.processedActionKeys.add(actionIdempotencyKey);
    }
    this.profile.xp = (this.profile.xp || 0) + amount;
    return this.profile.xp;
  }

  async recordDailyActivity(dateStr?: string): Promise<number> {
    const today = dateStr || new Date().toISOString().split('T')[0];

    if (this.recordedActivityDates.has(today)) {
      return this.profile.streakDays;
    }

    this.recordedActivityDates.add(today);

    const lastDate = this.profile.lastActiveDate;
    if (!lastDate) {
      this.profile.streakDays = 1;
    } else if (lastDate === today) {
      // Same day, streak unchanged
    } else {
      const prevMs = new Date(lastDate).getTime();
      const currMs = new Date(today).getTime();
      const diffDays = Math.round((currMs - prevMs) / (1000 * 60 * 60 * 24));

      if (diffDays === 1) {
        this.profile.streakDays = (this.profile.streakDays || 0) + 1;
      } else {
        this.profile.streakDays = 1;
      }
    }

    this.profile.lastActiveDate = today;
    return this.profile.streakDays;
  }

  async resetAll(): Promise<LearnerProfile> {
    this.profile = {
      ...DEFAULT_PROFILE,
      scoresByCourse: {},
      completedLessonIds: [],
      completedNodeIds: [],
      completedReinforcementIds: [],
    };
    this.answerAttempts = [];
    this.processedActionKeys.clear();
    this.recordedActivityDates.clear();
    return this.getProfile();
  }
}

export const inMemoryLearnerRepository = new InMemoryLearnerRepository();
